File: src/devtwin/core/health.py
"""Health scoring engine: combines detection, runtime, dependency, service,
and Git signals into one transparent score with concrete issues.

Scoring is intentionally simple and documented so it's testable and so no
single optional/missing tool can tank the whole score by itself: each issue
subtracts a fixed amount based on severity, floored at 0.
"""

from __future__ import annotations

from pathlib import Path

from devtwin.adapters import ADAPTERS, GENERIC_ADAPTER
from devtwin.adapters.base import version_mismatch
from devtwin.core.detector import detect_project
from devtwin.core.models import (
    AdapterResult,
    HealthIssue,
    HealthReport,
    Severity,
    Status,
)
from devtwin.core.workspace import inspect_workspace
from devtwin.services import detect_services

SEVERITY_PENALTY = {
    Severity.INFO: 0,
    Severity.LOW: 3,
    Severity.MEDIUM: 8,
    Severity.HIGH: 15,
    Severity.CRITICAL: 25,
}


def _dependency_names(adapter_results: list[AdapterResult]) -> set[str]:
    names: set[str] = set()
    for result in adapter_results:
        for note in result.dependencies.notes if result.dependencies else []:
            names.add(note)
    return names


def run_adapters(root: Path) -> list[AdapterResult]:
    results = [a.run(root) for a in ADAPTERS]
    applicable = [r for r in results if r.applicable]
    if not applicable:
        generic_result = GENERIC_ADAPTER.run(root)
        return [generic_result]
    return applicable


def _score(issues: list[HealthIssue]) -> int:
    score = 100
    for issue in issues:
        score -= SEVERITY_PENALTY.get(issue.severity, 0)
    return max(0, min(100, score))


def _status_for_score(score: int) -> Status:
    if score >= 90:
        return Status.OK
    if score >= 50:
        return Status.WARNING
    return Status.ERROR


def compute_health(root: str) -> HealthReport:
    profile = detect_project(root)
    workspace = inspect_workspace(root)
    path = Path(workspace.root)

    issues: list[HealthIssue] = []
    recommendations: list[str] = []

    if not workspace.exists:
        issues.append(
            HealthIssue(
                severity=Severity.CRITICAL,
                code="workspace.not_found",
                title="Workspace path does not exist",
                message=f"'{root}' could not be found on disk.",
                evidence=[f"path checked: {workspace.root}"],
                recommendation="Verify the workspace path passed to DevTwin.",
            )
        )
        return HealthReport(
            health_score=0,
            status=Status.ERROR,
            project_summary="Workspace not found",
            runtime_summary="unknown",
            service_summary="unknown",
            dependency_summary="unknown",
            issues=issues,
            recommendations=["Point DevTwin at a valid, existing directory."],
        )

    adapter_results = run_adapters(path)

    runtime_lines = []
    for result in adapter_results:
        issues.extend(result.issues)
        for rt in result.runtimes:
            runtime_lines.append(f"{rt.name}={rt.installed_version or rt.presence.value}")
            if (
                rt.required_version
                and rt.installed_version
                and version_mismatch(rt.required_version, rt.installed_version)
            ):
                issues.append(
                    HealthIssue(
                        severity=Severity.MEDIUM,
                        code=f"{result.ecosystem}.runtime_drift",
                        title=f"{rt.name} version differs from project requirement",
                        message=(
                            f"Project requires {rt.name} {rt.required_version} "
                            f"(from {rt.source}), but {rt.installed_version} is installed."
                        ),
                        evidence=[
                            f"required={rt.required_version}",
                            f"installed={rt.installed_version}",
                        ],
                        recommendation=f"Install/switch to {rt.name} {rt.required_version}.",
                    )
                )
        for dep_note in [] if not result.dependencies else result.dependencies.notes:
            if "not found" in dep_note or "may not be installed" in dep_note:
                issues.append(
                    HealthIssue(
                        severity=Severity.LOW,
                        code=f"{result.ecosystem}.dependencies_not_installed",
                        title="Dependencies may not be installed",
                        message=dep_note,
                        evidence=[dep_note],
                        recommendation="Run the ecosystem's install command before building or testing.",
                    )
                )

    dep_names = _dependency_names(adapter_results)
    services = detect_services(path, dep_names)
    for svc in services:
        if svc.required and not svc.running:
            issues.append(
                HealthIssue(
                    severity=Severity.HIGH,
                    code="service.not_running",
                    title=f"{svc.name} appears required but is not running",
                    message=(
                        f"Detected signals suggesting '{svc.name}' is required "
                        f"(confidence={svc.required_confidence}), but port {svc.port} is not listening."
                    ),
                    evidence=svc.source,
                    recommendation=f"Start {svc.name} (e.g. via docker compose or a local install).",
                    confidence=svc.required_confidence,
                )
            )

    if workspace.git and workspace.git.is_repo and workspace.git.has_conflicts:
        issues.append(
            HealthIssue(
                severity=Severity.HIGH,
                code="git.merge_conflicts",
                title="Unresolved merge conflicts",
                message="Git status shows unmerged paths.",
                evidence=[*workspace.git.staged_files, *workspace.git.unstaged_files][:10],
                recommendation="Resolve conflicts before continuing.",
            )
        )

    score = _score(issues)
    status = _status_for_score(score)

    recommendations.extend(sorted({i.recommendation for i in issues if i.recommendation}))

    ecosystems = ", ".join(profile.ecosystems) if profile.ecosystems else "unrecognized"
    project_summary = f"{profile.primary_ecosystem or 'generic'} project (detected: {ecosystems})"
    runtime_summary = ", ".join(runtime_lines) if runtime_lines else "no runtimes inspected"
    service_summary = (
        ", ".join(f"{s.name}={'running' if s.running else 'stopped'}" for s in services)
        if services
        else "no required services detected"
    )
    dep_summary_parts = []
    for result in adapter_results:
        if result.dependencies:
            dep_summary_parts.append(
                f"{result.ecosystem}:{result.dependencies.manager or 'unknown manager'}"
            )
    dependency_summary = (
        ", ".join(dep_summary_parts) if dep_summary_parts else "no dependency manifests found"
    )

    return HealthReport(
        health_score=score,
        status=status,
        project_summary=project_summary,
        runtime_summary=runtime_summary,
        service_summary=service_summary,
        dependency_summary=dependency_summary,
        issues=issues,
        recommendations=recommendations,
    )

File: mcphub/devtwin/src/core/health.ts
/**
 * Health scoring engine: combines detection, runtime, dependency, service,
 * and Git signals into one transparent score with concrete issues.
 *
 * Scoring is intentionally simple and documented so it's testable and so no
 * single optional/missing tool can tank the whole score by itself: each issue
 * subtracts a fixed amount based on severity, floored at 0.
 */

import { ADAPTERS, GENERIC_ADAPTER } from '../adapters/index.js';
import { versionMismatch } from '../adapters/base.js';
import { detectProject } from './detector.js';
import {
  Severity,
  Status,
  type AdapterResult,
  type HealthIssue,
  type HealthReport,
  type ServiceInfo,
} from './models.js';
import { inspectWorkspace } from './workspace.js';
import { detectServices } from '../services/index.js';

export const SEVERITY_PENALTY: Record<Severity, number> = {
  [Severity.INFO]: 0,
  [Severity.LOW]: 3,
  [Severity.MEDIUM]: 8,
  [Severity.HIGH]: 15,
  [Severity.CRITICAL]: 25,
};

function dependencyNames(adapterResults: AdapterResult[]): Set<string> {
  const names = new Set<string>();
  for (const result of adapterResults) {
    for (const note of result.dependencies ? result.dependencies.notes : []) {
      names.add(note);
    }
  }
  return names;
}

export async function runAdapters(root: string): Promise<AdapterResult[]> {
  // Adapters are independent: run them concurrently, but keep the registry's
  // order in the results so output stays deterministic.
  const results = await Promise.all(ADAPTERS.map((a) => a.run(root)));
  const applicable = results.filter((r) => r.applicable);
  if (applicable.length === 0) {
    const genericResult = await GENERIC_ADAPTER.run(root);
    return [genericResult];
  }
  return applicable;
}

function score(issues: HealthIssue[]): number {
  let value = 100;
  for (const issue of issues) {
    value -= SEVERITY_PENALTY[issue.severity] ?? 0;
  }
  return Math.max(0, Math.min(100, value));
}

function statusForScore(value: number): Status {
  if (value >= 90) return Status.OK;
  if (value >= 50) return Status.WARNING;
  return Status.ERROR;
}

export async function computeHealth(root: string): Promise<HealthReport> {
  const profile = detectProject(root);
  const workspace = await inspectWorkspace(root);
  const path = workspace.root;

  const issues: HealthIssue[] = [];
  const recommendations: string[] = [];

  if (!workspace.exists) {
    issues.push({
      severity: Severity.CRITICAL,
      code: 'workspace.not_found',
      title: 'Workspace path does not exist',
      message: `'${root}' could not be found on disk.`,
      evidence: [`path checked: ${workspace.root}`],
      recommendation: 'Verify the workspace path passed to DevTwin.',
      confidence: null,
    });
    return {
      health_score: 0,
      status: Status.ERROR,
      project_summary: 'Workspace not found',
      runtime_summary: 'unknown',
      service_summary: 'unknown',
      dependency_summary: 'unknown',
      issues,
      recommendations: ['Point DevTwin at a valid, existing directory.'],
    };
  }

  const adapterResults = await runAdapters(path);

  const runtimeLines: string[] = [];
  for (const result of adapterResults) {
    issues.push(...result.issues);
    for (const rt of result.runtimes) {
      runtimeLines.push(`${rt.name}=${rt.installed_version || rt.presence}`);
      if (
        rt.required_version &&
        rt.installed_version &&
        versionMismatch(rt.required_version, rt.installed_version)
      ) {
        issues.push({
          severity: Severity.MEDIUM,
          code: `${result.ecosystem}.runtime_drift`,
          title: `${rt.name} version differs from project requirement`,
          message:
            `Project requires ${rt.name} ${rt.required_version} ` +
            `(from ${rt.source}), but ${rt.installed_version} is installed.`,
          evidence: [
            `required=${rt.required_version}`,
            `installed=${rt.installed_version}`,
          ],
          recommendation: `Install/switch to ${rt.name} ${rt.required_version}.`,
          confidence: null,
        });
      }
    }
    for (const depNote of result.dependencies ? result.dependencies.notes : []) {
      if (depNote.includes('not found') || depNote.includes('may not be installed')) {
        issues.push({
          severity: Severity.LOW,
          code: `${result.ecosystem}.dependencies_not_installed`,
          title: 'Dependencies may not be installed',
          message: depNote,
          evidence: [depNote],
          recommendation: "Run the ecosystem's install command before building or testing.",
          confidence: null,
        });
      }
    }
  }

  const depNames = dependencyNames(adapterResults);
  let services: ServiceInfo[] = [];
  try {
    services = await detectServices(path, depNames);
  } catch {
    // Service detection is best-effort metadata; never let it break health.
    services = [];
  }
  for (const svc of services) {
    if (svc.required && !svc.running) {
      issues.push({
        severity: Severity.HIGH,
        code: 'service.not_running',
        title: `${svc.name} appears required but is not running`,
        message:
          `Detected signals suggesting '${svc.name}' is required ` +
          `(confidence=${svc.required_confidence}), but port ${svc.port} is not listening.`,
        evidence: svc.source,
        recommendation: `Start ${svc.name} (e.g. via docker compose or a local install).`,
        confidence: svc.required_confidence,
      });
    }
  }

  if (workspace.git && workspace.git.is_repo && workspace.git.has_conflicts) {
    issues.push({
      severity: Severity.HIGH,
      code: 'git.merge_conflicts',
      title: 'Unresolved merge conflicts',
      message: 'Git status shows unmerged paths.',
      evidence: [...workspace.git.staged_files, ...workspace.git.unstaged_files].slice(0, 10),
      recommendation: 'Resolve conflicts before continuing.',
      confidence: null,
    });
  }

  const healthScore = score(issues);
  const status = statusForScore(healthScore);

  const uniqueRecommendations = new Set<string>();
  for (const issue of issues) {
    if (issue.recommendation) uniqueRecommendations.add(issue.recommendation);
  }
  recommendations.push(...[...uniqueRecommendations].sort());

  const ecosystems = profile.ecosystems.length > 0 ? profile.ecosystems.join(', ') : 'unrecognized';
  const projectSummary = `${profile.primary_ecosystem || 'generic'} project (detected: ${ecosystems})`;
  const runtimeSummary = runtimeLines.length > 0 ? runtimeLines.join(', ') : 'no runtimes inspected';
  const serviceSummary =
    services.length > 0
      ? services.map((s) => `${s.name}=${s.running ? 'running' : 'stopped'}`).join(', ')
      : 'no required services detected';
  const depSummaryParts: string[] = [];
  for (const result of adapterResults) {
    if (result.dependencies) {
      depSummaryParts.push(
        `${result.ecosystem}:${result.dependencies.manager || 'unknown manager'}`
      );
    }
  }
  const dependencySummary =
    depSummaryParts.length > 0 ? depSummaryParts.join(', ') : 'no dependency manifests found';

  return {
    health_score: healthScore,
    status,
    project_summary: projectSummary,
    runtime_summary: runtimeSummary,
    service_summary: serviceSummary,
    dependency_summary: dependencySummary,
    issues,
    recommendations,
  };
}
